prefect_gcp/cloud_storage.py

```diff
@@ -753,6 +753,48 @@ async def list_blobs(self, folder: str = "") -> List["Blob"]:
 
         # Ignore folders
         return [blob for blob in blobs if not blob.name.endswith("/")]
+    
+    @sync_compatible
+    async def list_folders(self, folder: str = "") -> List[str]:
+        """
+        Lists all folders in the bucket.
+
+        Args:
+            folder: List all folders inside given folder.
+
+        Returns:
+            A list of folders.
+
+        Examples:
+            Get all folders from a bucket named "my-bucket".
+            ```python
+            from prefect_gcp.cloud_storage import GcsBucket
+
+            gcs_bucket = GcsBucket.load("my-bucket")
+            gcs_bucket.list_folders()
+            ```
+
+            Get all folders from a folder called years
+            ```python
+            from prefect_gcp.cloud_storage import GcsBucket
+
+            gcs_bucket = GcsBucket.load("my-bucket")
+            gcs_bucket.list_folders('years)
+            ```
+        """
+        client = self.gcp_credentials.get_cloud_storage_client()
+
+        bucket_path = self._join_bucket_folder(folder)
+        self.logger.info(f"Listing folders in bucket {bucket_path}.")
+        blobs = await run_sync_in_worker_thread(
+            client.list_blobs, self.bucket, prefix=bucket_path
+        )
+        blobs = list(blobs)
+        if bucket_path is not None:
+            unique_folders = list(set([blob.name.split(os.sep)[1] for blob in blobs if "." not in blob.name.split(os.sep)[1]]))
+            return [folder for folder in unique_folders if folder]
+        else:
+            return list(set([blob.name.split(os.sep)[0] for blob in blobs if blob.name.endswith(os.sep)]))
```

Comment:
I am having a hard time understanding this logic. Can you elaborate on the intent?

Is "." checking to see if it's a file?

Also, I don't think we should use os.sep since Windows will use "\" which isn't used in GCP.

Reply:
Yes, the set(list()) is for taking only the unique elements because there were folders showing up for each blob (e.g. ['subdirectory/file1.txt', 'subdirectory/file2.txt' ]  and the output would be subdirectory*2)

The '.' is indeed for files because these we're showing up if they were in the same directory as a subdirectory.

And the split()[0] and split()[1] is for showing the root directory or the 2nd level one

Reply:
We can replace os.sep with '/' as I didn't know that about Windows

Reply:
We can probably simplify by:
1. using a set comprehension directly
2. using PurePosixPath to handle discovery of folders
3. removing the if/else

Something like this:
```
from prefect_gcp.cloud_storage import GcsBucket
from pathlib import PurePosixPath

bucket = await GcsBucket.load("my-bucket")
blobs = client.list_blobs("my-sbx-bucket")
{str(PurePosixPath(blob.name).parent) for blob in blobs}
```